app/controllers/hiring_staff/vacancies_controller.rb

```diff
@@ -74,47 +67,46 @@ def redirect_if_published
     end
   end
 
+  def redirect_to_incomplete_step
+    return redirect_to organisation_job_build_path(@vacancy.id, :job_details) unless step_valid?(JobDetailsForm)
+    return redirect_to organisation_job_build_path(@vacancy.id, :pay_package) unless step_valid?(PayPackageForm)
+    return redirect_to organisation_job_build_path(@vacancy.id, :important_dates) unless step_valid?(ImportantDatesForm)
+    return redirect_to organisation_job_build_path(@vacancy.id, :supporting_documents) unless step_valid?(SupportingDocumentsForm)
+    return redirect_to organisation_job_build_path(@vacancy.id, :applying_for_the_job) unless step_valid?(ApplyingForTheJobForm)
+    return redirect_to organisation_job_build_path(@vacancy.id, :job_summary) unless step_valid?(JobSummaryForm)
+  end
+
   def redirect_unless_permitted
-    if @vacancy.state == "copy" && !@vacancy.valid?
+    if @vacancy.state == "copy" && !all_steps_valid?
       redirect_to organisation_job_review_path(@vacancy.id)
-    elsif @vacancy.state == "edit_published" && !@vacancy.valid?
+    elsif @vacancy.state == "edit_published" && !all_steps_valid?
       redirect_to edit_organisation_job_path(@vacancy.id)
-    elsif !@vacancy.valid?
+    elsif !all_steps_valid?
       redirect_to_incomplete_step
     end
   end
 
-  def redirect_to_incomplete_step
-    return redirect_to organisation_job_job_specification_path(@vacancy.id) unless step_valid?(JobSpecificationForm)
-    return redirect_to organisation_job_pay_package_path(@vacancy.id) unless step_valid?(PayPackageForm)
-    return redirect_to organisation_job_important_dates_path(@vacancy.id) unless step_valid?(ImportantDatesForm)
-    return redirect_to organisation_job_supporting_documents_path(@vacancy.id) unless
-      step_valid?(SupportingDocumentsForm)
-    return redirect_to organisation_job_application_details_path(@vacancy.id) unless step_valid?(ApplicationDetailsForm)
-    return redirect_to organisation_job_job_summary_path(@vacancy.id) unless step_valid?(JobSummaryForm)
-  end
-
-  def clear_cache_and_step
-    flash.clear
-    session[:current_step] = ""
-  end
-
   def set_completed_step
-    @vacancy.update(completed_step: current_step)
+    @vacancy.update(completed_step: current_step_number)
   end
 
   def update_vacancy_state
-    state = if params[:edit_draft] == "true" || @vacancy&.state == "edit"
+    state = if params[:edit_draft] == "true" || @vacancy.state == "edit"
```

Comment:
I know it's not part of this PR, but why is this not an enum? 😕

Reply:
Mainly because enums weren't used as much across the codebase when this was hacked together (for GA purposes), so there's no good reason it shouldn't be